### PR TITLE
application state added

### PR DESCRIPTION
**Pull request by tomervak**

### CODE REVIEW

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 1]
```diff
@@ -45,7 +46,20 @@ export class ApplicationsController {
...
+        return await this.applicationService.create(createApplicationDto, request.jwtPayLoad.sub);
+    }
+
+    @Post(':id')
+    @ApiOperation({ summary: 'add new state ' })
+    @ApiCreatedResponse({
+        description: 'applicationDto',
+        type: ApplicationDto
+    })
+    async createState(@Request() request: ApiRequest,
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts:58]: Change route and param name according to jira ticket https://groupy-group.atlassian.net/browse/JAT-10

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 2]
```diff
@@ -38,17 +40,21 @@ export class ApplicationsService {
...
         
         const savedApplication = await this.applicationRepository.create(newApplication);
         return ApplicationMapper.toApplicationDto(savedApplication);
 
     }
-    async update(id: string, updateApplicationDto: UpdateApplicationDto) {
+    async update(id: string, updateApplicationDto: UpdateApplicationDto, userId:string) {
         const application = await this.applicationRepository.findById(id);
-        if (!application) throw new NotFoundException('no such application');
+        if (!application || (application.userId !== userId)) throw new NotFoundException('no such application');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: Add admin check which was removed from the controller

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 3]
```diff
@@ -60,4 +66,20 @@ export class ApplicationsService {
...
+            throw new NotFoundException('no such application');
+        }
+        application.states.push(newState);
+        const updatedApplication={
+            companyName: application.companyName,
+            position: application.position,
+            states:application.states,
+            userId:userId
+        };
+        const updateApplication = await this.applicationRepository.update(appId, updatedApplication);
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: No need to create `updatedApplication`, just send `application` to `applicationRepository.update`

**Review by ido-rosenberger — CHANGES_REQUESTED** [review]

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 4]
```diff
@@ -45,7 +46,20 @@ export class ApplicationsController {
...
+    }
+
+    @Post(':applicationId/states')
+    @ApiOperation({ summary: 'add new state ' })
+    @ApiCreatedResponse({
+        description: 'applicationDto',
+        type: ApplicationDto
+    })
+    async createState(@Request() request: ApiRequest,
+        @Param('applicationId') id: string,
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: rename id param to applicationId

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 5]
```diff
@@ -27,28 +29,32 @@ export class ApplicationsService {
...
+            userId: userId
         }
-        
+
         const savedApplication = await this.applicationRepository.create(newApplication);
         return ApplicationMapper.toApplicationDto(savedApplication);
 
     }
-    async update(id: string, updateApplicationDto: UpdateApplicationDto) {
+    async update(id: string, updateApplicationDto: UpdateApplicationDto, userId: string, isAdmin: Boolean) {
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: use boolean istead of Boolean, Boolean is a JavaScript class

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 6]
```diff
@@ -27,28 +29,32 @@ export class ApplicationsService {
...
         const application = await this.applicationRepository.findById(id);
-        if (!application) throw new NotFoundException('no such application');
+        if (!application || !((application.userId !== userId) || isAdmin)) throw new NotFoundException('no such application');
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: De Morgan

**Review by ido-rosenberger — CHANGES_REQUESTED** [review]

File: job-application-tracker-api/src/applications/applications.controller.ts (hunk 1 above)
Comment (tomervak) [reply job-application-tracker-api/src/applications/applications.controller.ts:58]: lo hevanti ma ata roze

File: job-application-tracker-api/src/applications/applications.controller.ts [hunk 7]
```diff
@@ -58,11 +72,11 @@ export class ApplicationsController {
         @Param('id') id: string,
         @Body(ValidationPipe) updateApplicationDto: UpdateApplicationDto
     ) {
-        const userId=(await this.applicationService.findOne(updateApplicationDto.id)).userId;
-        if (request.jwtPayLoad.sub!==userId &&!request.jwtPayLoad.roles.includes(Role.admin)) {
-            throw new NotFoundException();
+        let isAdmin: Boolean = false;
+        if (request.jwtPayLoad.roles.includes(Role.admin)) {
+            isAdmin = true;
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.controller.ts]: let isAdmin: boolean = request.jwtPayLoad.roles.includes(Role.admin);

**Review by ido-rosenberger — APPROVED** [review]

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 8]
```diff
@@ -60,4 +66,14 @@ export class ApplicationsService {
...
     }
+
+    async createState(appId: string, newState: ApplicationState, userId: string) {
+        const application: Application = await this.applicationRepository.findById(appId);
+        if (!application || application.userId !== userId) {
+            throw new NotFoundException('no such application');
+        }
+        application.states.push(newState);
+        const updateApplication = await this.applicationRepository.update(appId, application);
+        return ApplicationMapper.toApplicationDto(updateApplication);
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts:77]: Only return the state dto

File: job-application-tracker-api/src/applications/applications.service.ts [hunk 9]
```diff
@@ -60,4 +66,14 @@ export class ApplicationsService {
         if (!deletedApplication) throw new NotFoundException('no such application');
         return ApplicationMapper.toApplicationDto(deletedApplication);
     }
+
+    async createState(appId: string, newState: ApplicationState, userId: string) {
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/applications/applications.service.ts]: newState: CreateStateDto

**Pull request merged**